server/src/services/maintenance.service.ts

```diff
@@ -0,0 +1,84 @@
+import { BadRequestException, Injectable } from '@nestjs/common';
+import { SignJWT } from 'jose';
+import { randomBytes } from 'node:crypto';
+import { OnEvent } from 'src/decorators';
+import { MaintenanceAuthDto } from 'src/dtos/maintenance.dto';
+import { SystemMetadataKey } from 'src/enum';
+import { BaseService } from 'src/services/base.service';
+import { MaintenanceModeState } from 'src/types';
+import { getExternalDomain } from 'src/utils/misc';
+
+/**
+ * This service is available outside of maintenance mode to manage maintenance mode
+ */
+@Injectable()
+export class MaintenanceService extends BaseService {
+  getMaintenanceMode(): Promise<MaintenanceModeState> {
+    return this.systemMetadataRepository
+      .get(SystemMetadataKey.MaintenanceMode)
+      .then((state) => state ?? { isMaintenanceMode: false });
+  }
+
+  login(): MaintenanceAuthDto {
+    throw new BadRequestException('Not in maintenance mode');
+  }
+
+  async startMaintenance(): Promise<{ secret: string }> {
+    const { isMaintenanceMode } = await this.getMaintenanceMode();
+    if (isMaintenanceMode) {
+      throw new BadRequestException('Already in maintenance mode');
+    }
+
+    const secret = MaintenanceService.generateSecret();
+    await this.systemMetadataRepository.set(SystemMetadataKey.MaintenanceMode, { isMaintenanceMode: true, secret });
+    await this.eventRepository.emit('AppRestart', { isMaintenanceMode: true });
+
+    return { secret };
+  }
+
+  endMaintenance(): void {
+    throw new BadRequestException('Not in maintenance mode');
+  }
+
+  @OnEvent({ name: 'AppRestart', server: true })
+  onRestart(): void {
+    this.maintenanceRepository.exitApp();
+  }
+
+  async createLoginUrl(auth: MaintenanceAuthDto, secret?: string): Promise<string> {
+    const { server } = await this.getConfig({ withCache: true });
+    const baseUrl = getExternalDomain(server);
+
+    secret ??= await this.getMaintenanceMode().then((state) => {
+      if (!state.isMaintenanceMode) {
+        throw new Error('Not in maintenance mode');
+      }
+
+      return state.secret;
+    });
```

Comment:
I would personally move this block inside the .then out of there (and not try to cram it into a single-line assignment. So more of like

```ts
const { isMaintenanceMode, secret: currentSecret } = await this.getMaintenanceMode();

if (!isMaintenanceMode) {
	throw new Error(...)
} 

return await MaintenanceService.createLoginUrl(baseUrl, auth, secret ?? currentSecret);
```

Reply:
Refactored out but instead reassigning `secret` as I'm only running the state fetch if `secret` is not set.